### PR TITLE
[#45] - Support cli options from env variables and yaml config file

### PR DESCRIPTION
**Pull request by usmansaleem**
Fix issues #45 and #46.

Read default values from config file specified using `--config-file` or environment variables prefixed with `ETH2SIGNER_`. The precedence order is cli, env, config file

### CODE REVIEW

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/ConfigFileCommand.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import picocli.CommandLine.Command;
+import picocli.CommandLine.Option;
+import picocli.CommandLine.Unmatched;
+
+@Command(mixinStandardHelpOptions = true)
+class ConfigFileCommand {
+  @Option(names = CONFIG_FILE_OPTION_NAME, description = "...")
+  File configPath = null;
+
+  @Unmatched List<String> unmatched;
```
Comment (jframe) [inline commandline/src/main/java/tech/pegasys/eth2signer/commandline/ConfigFileCommand.java:29]: why is this needed? doesn't look it is ever used?

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/ConfigFileCommand.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import java.io.File;
+import java.util.List;
+
+import picocli.CommandLine.Command;
+import picocli.CommandLine.Option;
+import picocli.CommandLine.Unmatched;
+
+@Command(mixinStandardHelpOptions = true)
+class ConfigFileCommand {
+  @Option(names = CONFIG_FILE_OPTION_NAME, description = "...")
```
Comment (jframe) [inline commandline/src/main/java/tech/pegasys/eth2signer/commandline/ConfigFileCommand.java:26]: can a description be added

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/CascadingDefaultProvider.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+import picocli.CommandLine.IDefaultValueProvider;
+import picocli.CommandLine.Model.ArgSpec;
+
+public class CascadingDefaultProvider implements IDefaultValueProvider {
+  private final IDefaultValueProvider defaultValueProvider1;
+  private final IDefaultValueProvider defaultValueProvider2;
+
+  public CascadingDefaultProvider(
+      final IDefaultValueProvider defaultValueProvider1,
```
Comment (jframe) [inline commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/CascadingDefaultProvider.java:23]: Why limit this to just two defaultProviders? The version in Besu can handle multiple and the code is simpler

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProvider.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+  @Override
+  public String defaultValue(final ArgSpec argSpec) {
+    loadConfigurationFromFile();
+
+    // only options can be used in config because a name is needed for the key
+    // so we skip default for positional params
+    return argSpec.isOption() ? getConfigurationValue(((OptionSpec) argSpec)) : null;
+  }
+
+  private void loadConfigurationFromFile() {
```
Comment (jframe) [inline commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProvider.java:57]: seems a bit odd that the function doesn't return a value since it is loading something. Thinking it should return the Map<String, Object>. I don't think the stateful field variable result is needed at all

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProvider.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+    final Set<String> unknownOptionsList =
+        result.keySet().stream()
+            .filter(option -> !commandSpec.optionsMap().containsKey("--" + option))
+            .collect(Collectors.toSet());
+
+    if (!unknownOptionsList.isEmpty()) {
+      final String options = unknownOptionsList.size() > 1 ? "options" : "option";
+      final String csvUnknownOptions =
+          unknownOptionsList.stream().collect(Collectors.joining(", "));
```
Comment (jframe) [inline commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProvider.java:101]: can be replaced with String.join(", ", unknownOptionsList);

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/CommandlineParser.java [hunk 6]
```diff
@@ -12,41 +12,79 @@
...
+      final Map<String, String> environment) {
     this.baseCommand = baseCommand;
     this.outputWriter = outputWriter;
     this.errorWriter = errorWriter;
+    this.environment = environment;
   }
 
   public int parseCommandLine(final String... args) {
+    // first pass to obtain config file if specified
+    ConfigFileCommand configFileCommand = new ConfigFileCommand();
```
Comment (jframe) [inline commandline/src/main/java/tech/pegasys/eth2signer/commandline/CommandlineParser.java:50]: nit: final

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/CommandlineParser.java [hunk 7]
```diff
@@ -12,41 +12,79 @@
...
+    final CommandLine configFileCommandLine = new CommandLine(configFileCommand);
+    configFileCommandLine.parseArgs(args);
+    if (configFileCommandLine.isUsageHelpRequested()) {
+      new CommandLine(baseCommand).usage(outputWriter);
+      return 0;
```
Comment (jframe) [inline commandline/src/main/java/tech/pegasys/eth2signer/commandline/CommandlineParser.java:55]: Nicer to use commandLine.getCommandSpec().exitCodeOnUsageHelp

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/CommandlineParser.java [hunk 8]
```diff
@@ -12,41 +12,79 @@
...
+    } else if (configFileCommandLine.isVersionHelpRequested()) {
+      new CommandLine(baseCommand).printVersionHelp(outputWriter);
+      return 0;
```
Comment (jframe) [inline commandline/src/main/java/tech/pegasys/eth2signer/commandline/CommandlineParser.java:58]: Nicer to use commandLine.getCommandSpec().exitCodeOnVersionHelp

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/ConfigFileCommand.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+import java.io.File;
+import java.util.List;
+
+import picocli.CommandLine.Command;
+import picocli.CommandLine.Option;
+import picocli.CommandLine.Unmatched;
+
+@Command(mixinStandardHelpOptions = true)
+class ConfigFileCommand {
```
Comment (jframe) [inline commandline/src/main/java/tech/pegasys/eth2signer/commandline/ConfigFileCommand.java:25]: Found this a bit confusing at first as this isn't the real config file command just there for the initial parse. Is there a better name can be used for this or perhaps just make this an internal class for the CommandLineParser?

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/Eth2SignerCommand.java [hunk 10]
```diff
@@ -58,6 +61,13 @@
 
   private static final Logger LOG = LogManager.getLogger();
 
+  @SuppressWarnings("UnusedVariable")
+  @CommandLine.Option(
```
Comment (jframe) [inline commandline/src/main/java/tech/pegasys/eth2signer/commandline/Eth2SignerCommand.java:65]: Could this be done as a mixin perhaps, so we could use the same command for both the initial and base command?

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProvider.java [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    if (result != null) {
+      return;
+    }
+
+    try {
+      final String configYaml = Files.readString(configFile.toPath());
+      final Map<String, Object> result = new Yaml().load(configYaml);
+      checkConfigurationValidity(result == null || result.isEmpty());
+      checkUnknownOptions(result);
+      this.result = result;
```
Comment (jframe) [inline commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProvider.java:67]: Would be nicer to just return the result instead of assigning to a field variable

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProvider.java [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+  public String defaultValue(final ArgSpec argSpec) {
+    loadConfigurationFromFile();
+
+    // only options can be used in config because a name is needed for the key
+    // so we skip default for positional params
+    return argSpec.isOption() ? getConfigurationValue(((OptionSpec) argSpec)) : null;
+  }
+
+  private void loadConfigurationFromFile() {
+    if (result != null) {
```
Comment (jframe) [inline commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProvider.java:58]: Why is this needed?

File: commandline/src/test/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProviderTest.java [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+    final String extraYamlOptions = "";
+    final File configFile =
+        Files.writeString(tempDir.resolve("config.yaml"), extraYamlOptions).toFile();
+    final Eth2SignerCommand eth2SignerCommand = new Eth2SignerCommand();
+    final CommandLine commandLine = new CommandLine(eth2SignerCommand);
+    commandLine.registerConverter(Level.class, Level::valueOf);
+    commandLine.setDefaultValueProvider(new YamlConfigFileDefaultProvider(commandLine, configFile));
+
+    final String[] args = CmdlineHelpers.validBaseCommandOptions().split(" ");
+    assertThatExceptionOfType(CommandLine.ParameterException.class)
```
Comment (jframe) [inline commandline/src/test/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProviderTest.java:101]: Do we care if the config file is empty? Since we have default options for everything having an empty config could be considered valid?

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/CommandlineParser.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/eth2signer/commandline/CommandlineParser.java]: 👍

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/CommandlineParser.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/eth2signer/commandline/CommandlineParser.java]: done, extracted out separate method as well.

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/CommandlineParser.java [hunk 14]
```diff
@@ -12,41 +12,79 @@
...
     return commandLine.execute(args);
   }
 
+  private IDefaultValueProvider defaultValueProvider(
+      final CommandLine commandLine, final Optional<File> configFile) {
+    final IDefaultValueProvider defaultValueProvider;
+    if (configFile.isPresent()) {
+      defaultValueProvider =
+          new CascadingDefaultProvider(
+              new EnvironmentVariableDefaultProvider(environment),
```
[the remark this answers is not in the document]
Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/eth2signer/commandline/CommandlineParser.java:90]: If config file is specified, both env variables provider and config file provider will be used (with env have higher precedence than config file). Else only env variables provider will be used.

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/ConfigFileCommand.java (hunk 1 above)
Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/eth2signer/commandline/ConfigFileCommand.java:29]: This is the "two pass" approach. We first create `commandLine(ConfigFileCommand)` which allows us to read optional "config file" option, rest of the command options will be treated as unmatched. This will avoid invoking any option validations by picocli which would be done in the second pass, which is against `Eth2SignerCommand`.

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/ConfigFileCommand.java (hunk 2 above)
Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/eth2signer/commandline/ConfigFileCommand.java:26]: We don't print usage from "ConfigFileCommand". It is purely used as a way to parse config file option in the "first pass". The actual description comes in Eth2SignerCommand's configFile option which gets printed when user ask from usage help.

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProvider.java (hunk 4 above)
Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProvider.java:57]: It was copied from Besu. This is only required to be executed once during parsing (not initial object creation). I'll modify it slightly to avoid modify instance variable.

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/ConfigFileCommand.java (hunk 9 above)
Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/eth2signer/commandline/ConfigFileCommand.java:25]: It was originally internal command ... I think I should move it back.

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProvider.java (hunk 12 above)
Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProvider.java:58]: The section for "extra invalid options" require the configuration file to be initialized during parse method as the command line options are fully parsed and initialised. During the initial object creation, we don't have commandLine.parse yet.

Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProvider.java:58]: Or may be we can perform "invalid options" check once in parse method and load configuration file when we create object?

File: commandline/src/test/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProviderTest.java (hunk 13 above)
Comment (usmansaleem) [reply commandline/src/test/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProviderTest.java:101]: The check was there in Besu's "toml" configuration. This implementation is roughly based on that.

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProvider.java (hunk 11 above)
Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProvider.java]: we need the "result" in consecutive calls of "defaultValue".

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProvider.java (hunk 5 above)
Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProvider.java]: 👍 Updated test cases as well.

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/CascadingDefaultProvider.java (hunk 3 above)
Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/CascadingDefaultProvider.java]: refactored.

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/ConfigFileCommand.java (hunk 2 above)
Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/eth2signer/commandline/ConfigFileCommand.java]: Based on https://github.com/remkop/picocli/blob/master/picocli-examples/src/main/java/picocli/examples/defaultprovider/TwoPass.java

File: commandline/src/main/java/tech/pegasys/eth2signer/commandline/CommandlineParser.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/eth2signer/commandline/CommandlineParser.java]: slightly refactored to make it more readable.
